src/hotspot/share/cds/filemap.cpp

```diff
@@ -2453,6 +2454,24 @@ bool FileMapHeader::validate() {
     return false;
   }
 
+  if (is_static()) {
+    const char* err = nullptr;
+    if (CDSConfig::is_valhalla_preview()) {
+      if (!_has_valhalla_patched_classes) {
+        err = "not created";
+      }
+    } else {
+      if (_has_valhalla_patched_classes) {
+        err = "created";
+      }
+    }
+    if (err != nullptr) {
+      log_warning(cds)("This archive was %s with --enable-preview -XX:+EnableValhalla. It is "
+                         "incompatible with the current JVM setting", err);
+      return false;
+    }
+  }
+
```

Comment:
Another VM option, `InlineFieldMaxFlatSize`, can influence the field layout, and by consequence, the validity of the `InstanceKlass` from the archive. It is probably needed to encode the value of `InlineFieldMaxFlatSize` in the archive to be able to check against the current VM options. We can decide to completely reject the archive if `InlineFieldMaxFlatSize` options don't match, or we can have a per class approach, as most classes won't have flat fields, and only reject them with a field layout not compliant with the current flag value.

Reply:
We have such a check already (see [JDK-8272290](https://bugs.openjdk.org/browse/JDK-8272290)]